### PR TITLE
Abstract WAN endpoints

### PR DESCRIPTION
**Pull request by emrahkocaman**
- Configuration changes for Abstract WAN endpoints.
- Schema updates to 3.7

### CODE REVIEW

**Comment by emrahkocaman** [comment]
run-lab-run

**Comment by emrahkocaman** [comment]
Unrelated test failures.

image

**Comment by emrahkocaman** [comment]
verify

**Comment by emrahkocaman** [comment]
Yet another unrelated failure

[com.hazelcast.replicatedmap.ReplicatedMapHitsAndLastAccessTimeTest.test_hitsAndLastAccessTimeAreSet_with2Nodes_object] (https://hazelcast-l337.ci.cloudbees.com/job/Hazelcast-3.x-pr-builder/com.hazelcast$hazelcast/17986/testReport/junit/com.hazelcast.replicatedmap/ReplicatedMapHitsAndLastAccessTimeTest/test_hitsAndLastAccessTimeAreSet_with2Nodes_object/)

**Comment by emrahkocaman** [comment]
run-lab-run

File: hazelcast-spring/src/main/java/com/hazelcast/spring/HazelcastConfigBeanDefinitionParser.java [hunk 1]
```diff
@@ -833,9 +833,55 @@ public void handleWanReplication(Node node) {
...
+                    final NamedNodeMap attrs = n.getAttributes();
+                    Node classNameNode = attrs.getNamedItem("class-name");
+                    String className = classNameNode != null ? getTextContent(classNameNode) : null;
+                    Node implNode = attrs.getNamedItem("implementation");
+                    String implementation = implNode != null ? getTextContent(implNode) : null;
+                    publisherConfigBuilder.addPropertyValue("className", className);
+                    if (implementation != null) {
+                        publisherConfigBuilder.addPropertyReference("implementation", implementation);
+                    }
+                    Assert.isTrue(className != null || implementation != null, "One of 'class-name' or 'implementation' "
```
Comment (serkan-ozal) [inline hazelcast-spring/src/main/java/com/hazelcast/spring/HazelcastConfigBeanDefinitionParser.java]: what is the expected behaviour when both of the `className` and `implementation` are specified?

File: hazelcast-spring/src/main/java/com/hazelcast/spring/HazelcastConfigBeanDefinitionParser.java [hunk 2]
```diff
@@ -833,9 +833,55 @@ public void handleWanReplication(Node node) {
...
+                } else if ("wan-consumer".equals(nName)) {
+                    final BeanDefinitionBuilder consumerConfigBuilder = createBeanBuilder(WanConsumerConfig.class);
+                    final NamedNodeMap attrs = n.getAttributes();
+                    Node classNameNode = attrs.getNamedItem("class-name");
+                    String className = classNameNode != null ? getTextContent(classNameNode) : null;
+                    Node implNode = attrs.getNamedItem("implementation");
+                    String implementation = implNode != null ? getTextContent(implNode) : null;
+                    consumerConfigBuilder.addPropertyValue("className", className);
+                    if (implementation != null) {
+                        consumerConfigBuilder.addPropertyReference("implementation", implementation);
```
Comment (serkan-ozal) [inline hazelcast-spring/src/main/java/com/hazelcast/spring/HazelcastConfigBeanDefinitionParser.java]: just for curiosity, what is the difference between `addPropertyValue` and `addPropertyReference` in here?

File: hazelcast/src/main/java/com/hazelcast/config/AbstractWanTargetConfig.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+package com.hazelcast.config;
+
+/**
+ * Contains common configurations for WAN publishers.
+ */
+public abstract class AbstractWanTargetConfig implements IWanPublisherConfig {
+
+    private static final int DEFAULT_QUEUE_CAPACITY = 10000;
+    private static final WANQueueFullBehavior QUEUE_FULL_BEHAVIOR = WANQueueFullBehavior.DISCARD_AFTER_MUTATION;
```
Comment (serkan-ozal) [inline hazelcast/src/main/java/com/hazelcast/config/AbstractWanTargetConfig.java]: might be `DEFAULT_QUEUE_FULL_BEHAVIOR`

File: hazelcast/src/main/java/com/hazelcast/config/WanConsumerConfig.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+import java.util.Map;
+
+/**
+ * Config to be used by WanReplicationConsumer instances. (EE only)
+ */
+public class WanConsumerConfig {
+
+    private Map<String, Comparable> properties = new HashMap<String, Comparable>();
+    private String className;
+    private Object implementation;
```
Comment (serkan-ozal) [inline hazelcast/src/main/java/com/hazelcast/config/WanConsumerConfig.java]: Will be there a contract point (interface) for all consumer implementations?

**Comment by serkan-ozal** [comment]
:+1: minor comments

File: hazelcast/src/main/java/com/hazelcast/config/WanConsumerConfig.java (hunk 4 above)
Comment (emrahkocaman) [inline hazelcast/src/main/java/com/hazelcast/config/WanConsumerConfig.java]: Yeap, please see https://github.com/hazelcast/hazelcast-enterprise/issues/815

File: hazelcast-spring/src/main/java/com/hazelcast/spring/HazelcastConfigBeanDefinitionParser.java (hunk 2 above)
Comment (emrahkocaman) [inline hazelcast-spring/src/main/java/com/hazelcast/spring/HazelcastConfigBeanDefinitionParser.java]: `addPropertyValue` is used for primitive types. `addPropertyReference` is used to inject an existing Spring bean.

File: hazelcast-spring/src/main/java/com/hazelcast/spring/HazelcastConfigBeanDefinitionParser.java (hunk 1 above)
Comment (emrahkocaman) [inline hazelcast-spring/src/main/java/com/hazelcast/spring/HazelcastConfigBeanDefinitionParser.java]: This is decided in `WanReplicationService`. "class-name" is ignored in that case.

File: hazelcast/src/main/java/com/hazelcast/config/WanConsumerConfig.java (hunk 4 above)
Comment (serkan-ozal) [inline hazelcast/src/main/java/com/hazelcast/config/WanConsumerConfig.java]: is there any reason for keeping this interface at EE?

File: hazelcast-spring/src/main/java/com/hazelcast/spring/HazelcastConfigBeanDefinitionParser.java (hunk 2 above)
Comment (serkan-ozal) [inline hazelcast-spring/src/main/java/com/hazelcast/spring/HazelcastConfigBeanDefinitionParser.java]: i see thanks

File: hazelcast-spring/src/test/resources/com/hazelcast/spring/fullConfig-applicationContext-hazelcast.xml [hunk 5]
```diff
@@ -49,8 +49,8 @@
                 <hz:property name="hazelcast.merge.first.run.delay.seconds">5</hz:property>
                 <hz:property name="hazelcast.merge.next.run.delay.seconds">5</hz:property>
             </hz:properties>
-            <hz:wan-replication name="testWan" snapshot-enabled="false">
-                <hz:target-cluster group-name="tokyo" group-password="tokyo-pass">
+            <hz:wan-replication name="testWan">
+                <hz:target-cluster group-name="tokyo" group-password="tokyo-pass" snapshot-enabled="false">
```
Comment (mesutcelik) [inline hazelcast-spring/src/test/resources/com/hazelcast/spring/fullConfig-applicationContext-hazelcast.xml]: Can you please ensure that this is mentioned in the Release Notes? That requires XML config change when upgraded to 3.7.

please also fail fast when snapshot-enabled=true but Batch Publisher is not used.

File: hazelcast-spring/src/main/java/com/hazelcast/spring/HazelcastConfigBeanDefinitionParser.java (hunk 1 above)
Comment (mesutcelik) [inline hazelcast-spring/src/main/java/com/hazelcast/spring/HazelcastConfigBeanDefinitionParser.java]: @emrahkocaman , 
please check if we can throw InvalidConfig when both is defined. @bilalyasar ?

File: hazelcast-spring/src/test/resources/com/hazelcast/spring/cache/jCacheClientCacheManager-applicationContext-hazelcast.xml [hunk 6]
```diff
@@ -24,7 +24,7 @@
 		http://www.springframework.org/schema/cache
 		http://www.springframework.org/schema/cache/spring-cache.xsd
 		http://www.hazelcast.com/schema/spring
-		http://www.hazelcast.com/schema/spring/hazelcast-spring-3.6.xsd">
+		http://www.hazelcast.com/schema/spring/hazelcast-spring-3.7.xsd">
```
Comment (mesutcelik) [inline hazelcast-spring/src/test/resources/com/hazelcast/spring/cache/jCacheClientCacheManager-applicationContext-hazelcast.xml]: I know this is annoying but we have an issue for that. I hope we get it done soon.

File: hazelcast/src/main/java/com/hazelcast/config/IWanPublisherConfig.java [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.config;
+
+/**
+ * Defines common attributes for both custom and provided Wan replication implementations.
```
Comment (mesutcelik) [inline hazelcast/src/main/java/com/hazelcast/config/IWanPublisherConfig.java]: WAN Replication in uppercases.

File: hazelcast/src/main/resources/hazelcast-config-3.7.xsd [hunk 8]
```diff
@@ -27,12 +27,12 @@
                 <xs:element ref="import"/>
                 <xs:element name="group" type="cluster-group" minOccurs="0" maxOccurs="1"/>
                 <xs:element name="license-key" type="xs:string" minOccurs="0" maxOccurs="1">
-	                <xs:annotation>
-	                    <xs:documentation>
-	                        To use Hazelcast Enterprise, you need to set the license key here or programmatically.
-	                    </xs:documentation>
-	                </xs:annotation>
-	            </xs:element>
+                    <xs:annotation>
```
Comment (mesutcelik) [inline hazelcast/src/main/resources/hazelcast-config-3.7.xsd]: so many changes in `hazelcast-config-3.7.xsd`?

**Comment by mesutcelik** [comment]
:+1:  with minor comments

File: hazelcast/src/main/java/com/hazelcast/config/IWanPublisherConfig.java [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+ */
+public interface IWanPublisherConfig {
```
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/config/IWanPublisherConfig.java]: Why the I prefix?

File: hazelcast/src/main/java/com/hazelcast/config/WanReplicationConfig.java [hunk 10]
```diff
@@ -24,22 +24,15 @@
...
 
     String name;
-    List<WanTargetClusterConfig> targetClusterConfigs;
-
-    /**
-     * This property is only valid when used with WAN Batch replication, Enterprise Only
-     * When enabled, only the latest {@link com.hazelcast.wan.WanReplicationEvent} of a key is sent to target
-     */
-    boolean snapshotEnabled;
+    List<WanTargetClusterConfig> targetClusterConfigs = new ArrayList<WanTargetClusterConfig>(2);
```
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/config/WanReplicationConfig.java]: No access modifiers for the fields?

File: hazelcast/src/main/java/com/hazelcast/config/WanReplicationConfig.java [hunk 11]
```diff
@@ -58,19 +51,33 @@ public WanReplicationConfig setName(String name) {
...
+    }
+
+    public List<WanPublisherConfig> getWanPublisherConfigs() {
+        return wanPublisherConfigs;
     }
 
-    public void setSnapshotEnabled(boolean snapshotEnabled) {
-        this.snapshotEnabled = snapshotEnabled;
+    public WanReplicationConfig setWanPublisherConfigs(List<WanPublisherConfig> wanPublisherConfigs) {
+        this.wanPublisherConfigs = wanPublisherConfigs;
```
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/config/WanReplicationConfig.java]: Do you want to allow for a null list to be inserted?

File: hazelcast/src/main/java/com/hazelcast/config/IWanPublisherConfig.java (hunk 9 above)
Comment (emrahkocaman) [inline hazelcast/src/main/java/com/hazelcast/config/IWanPublisherConfig.java]: Interface is no more needed, removed.

File: hazelcast/src/main/java/com/hazelcast/config/WanReplicationConfig.java (hunk 10 above)
Comment (emrahkocaman) [inline hazelcast/src/main/java/com/hazelcast/config/WanReplicationConfig.java]: Added.

File: hazelcast/src/main/java/com/hazelcast/config/WanReplicationConfig.java (hunk 11 above)
Comment (emrahkocaman) [inline hazelcast/src/main/java/com/hazelcast/config/WanReplicationConfig.java]: Check added.

File: hazelcast/src/main/java/com/hazelcast/config/WanPublisherConfig.java [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+        this.properties = properties;
+    }
+
+    @Override
+    public String toString() {
+        return "WanPublisherConfig{"
+                + "properties=" + properties
+                + ", className='" + className + '\''
+                + ", implementation=" + implementation
+                + "  groupName='" + groupName + '\''
```
Comment (bilalyasar) [inline hazelcast/src/main/java/com/hazelcast/config/WanPublisherConfig.java]: missing comma

File: hazelcast/src/main/java/com/hazelcast/config/WanConsumerConfig.java [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+        return implementation;
+    }
+
+    public WanConsumerConfig setImplementation(Object implementation) {
+        this.implementation = implementation;
+        return this;
+    }
+
+    public void setProperties(Map<String, Comparable> properties) {
+        this.properties = properties;
```
Comment (bilalyasar) [inline hazelcast/src/main/java/com/hazelcast/config/WanConsumerConfig.java]: missing `return this;`?

File: hazelcast/src/main/java/com/hazelcast/config/WanPublisherConfig.java [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+    private Map<String, Comparable> properties = new HashMap<String, Comparable>();
+    private String className;
+    private Object implementation;
+
+    public String getGroupName() {
+        return groupName;
+    }
+
+    public void setGroupName(String groupName) {
+        this.groupName = groupName;
```
Comment (bilalyasar) [inline hazelcast/src/main/java/com/hazelcast/config/WanPublisherConfig.java]: i think all setters should have `return this;`

File: hazelcast/src/main/java/com/hazelcast/config/WanPublisherConfig.java (hunk 12 above)
Comment (emrahkocaman) [inline hazelcast/src/main/java/com/hazelcast/config/WanPublisherConfig.java]: :+1:

File: hazelcast/src/main/java/com/hazelcast/config/WanConsumerConfig.java (hunk 13 above)
Comment (emrahkocaman) [inline hazelcast/src/main/java/com/hazelcast/config/WanConsumerConfig.java]: :+1:

File: hazelcast/src/main/java/com/hazelcast/config/WanPublisherConfig.java (hunk 14 above)
Comment (emrahkocaman) [inline hazelcast/src/main/java/com/hazelcast/config/WanPublisherConfig.java]: :+1:

**Comment by bilalyasar** [comment]
:+1:

**Pull request merged**